### ISSUE TITLE
criteria in prefix_

### ISSUE THREAD

**Issue by rja172** [skorch-dev/skorch#916]
Hi all,

I am working on multi-task training problem and writing my own weighted cross entropy loss function. I need to pass weights for each head. I was able to add `criterion__task_weights` and skorch properly initialized my loss function. However during validate_params, it fails because 'criterion' is not one of elements of `NeuralNet.prefix_`. For now I was able to over-write and get it working but wondering **is there any particular reason why `criterion` is not included in the list of `Neuralnet.prefix_`?**

**Comment by BenjaminBossan** [comment]
There is no `NeuralNet.prefix_` attribute, I assume you mean `NeuralNet.prefixes_`. The `prefixes_` for the module and criterion are populated automatically when the module and criterion are initialized. If it's not present, my first guess would be that you have some custom initialization code for the criterion which somehow bypasses the automatic addition. Is that possible?

If this is indeed the case, please check [this description](https://skorch.readthedocs.io/en/stable/user/customization.html#initialization-and-custom-modules) of how to correctly initialize custom modules and criteria.

If this is not the case, it could be a bug. Could you please paste some code to reproduce the issue?

**Comment by BenjaminBossan** [comment]
@rja172 any updates?

**Comment by rja172** [comment]
Hi @BenjaminBossan -- You are right. I do have a custom criterion that requires some initialization parameters. However, as I found out the problem is that `NeuralNet.prefixes_` doesn't include `criterion` as one of the list items and hence custom initialization of criterion fails. Below is the code that fixed my issue
```
from skorch import NeuralNet
# added 'criterion' to the last
NeuralNet.prefixes_ = ['iterator_train', 'iterator_valid', 'callbacks', 'dataset', 'criterion']
```

**Comment by BenjaminBossan** [comment]
Thanks for getting back. 

> However, as I found out the problem is that `NeuralNet.prefixes_` doesn't include `criterion` as one of the list items and hence custom initialization of criterion fails. Below is the code that fixed my issue

What you observed is actually a feature. `criterion` and `module` are not in this list because they are added _dynamically_ during initialization. So e.g. if you do:

```python
class MyNet(NeuralNet):
    def initialize_criterion(self):
        super().initialize_criterion()  # initializes net.criterion_

        # add an additional criterion called 'other_criterion_'
        params = self.get_params_for('other_criterion')
        self.other_criterion_ = nn.BCELoss(**params)
        return self
```

Then `criterion` and `other_criterion` are added automatically to `prefixes_`. For this to work, however, your code must be inside of `initialize_criterion`. I suspect that you add your criterion differently, which is why its prefix is not automatically added. Is that possible?

Anyway, if your code works with your fix and there are no other issues, feel free to close the issue. I just wanted to clarify what the "canonical" method is to add custom criteria in skorch.

**Comment by BenjaminBossan** [comment]
Since there was no further reply, I assume the issue is resolved. Feel free to re-open if not.
